x/gamm/keeper/grpc_query.go

```diff
@@ -140,6 +140,59 @@ func (q Querier) PoolType(ctx context.Context, req *types.QueryPoolTypeRequest)
 	}, err
 }
 
+// CalcJoinPoolShares queries the amount of shares you get by providing specific amount of tokens
+func (q Querier) CalcJoinPoolShares(ctx context.Context, req *types.QueryCalcJoinPoolSharesRequest) (*types.QueryCalcJoinPoolSharesResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "empty request")
+	}
+	if req.TokensIn == nil {
+		return nil, status.Error(codes.InvalidArgument, "no tokens in")
+	}
+
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	pool, err := q.Keeper.getPoolForSwap(sdkCtx, req.PoolId)
```

Comment:
For balancer pools, this method may call `PokePool()` which updates pool weights. Is this acceptable?

CC: @ValarDragon

Reply:
Yeah this is fine, because:
* this won't get saved
* Even if it did, cosmwasm would revert the result